Delete and edit custom commands history items

stefanhaller opened the pull request:
- **PR Description**

Allow deleting and editing custom command history items. Deleting is done by hitting `d` on a suggestion; editing is done by hitting `e`, which fills the selected item into the command prompt for further editing.

Closes #2528.

- **Please check if the PR fulfills these requirements**

* [x] Cheatsheets are up-to-date (run `go generate ./...`)
* [x] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [x] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [x] Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))
* [ ] Docs (specifically `docs/Config.md`) have been updated if necessary
* [x] You've read through your own file changes for silly mistakes etc

stefanhaller commented:
Marking this as a draft for now because I didn't add tests yet. Before I spend time on that I first wanted to confirm that we want this. (I find it pretty useful myself.)

mark2185 commented:
Just for future reference, this will close #2528 .

jesseduffield commented on pkg/gui/controllers/suggestions_controller.go:
```diff
@@ -40,8 +40,24 @@ func (self *SuggestionsController) GetKeybindings(opts types.KeybindingsOpts) []
-			Handler: func() error { return self.c.ReplaceContext(self.c.Contexts().Confirmation) },
+			Key: opts.GetKey(opts.Config.Universal.TogglePanel),
+			Handler: func() error {
+				if self.context().State.FillInTextOnTogglePanel {
```
If this returns false, we should still replace the context, right?

stefanhaller commented:
@jesseduffield I changed the keybinding for editing a history command from tab to "e", as discussed; and added a commit that shows the keybindings in the subtitle of the suggestions panel (only if it is focused).

stefanhaller commented on pkg/gui/controllers/confirmation_controller.go:
```diff
@@ -39,6 +41,14 @@ func (self *ConfirmationController) GetKeybindings(opts types.KeybindingsOpts) [
+						// are also editable, so we show both keybindings
+						subtitle = fmt.Sprintf(self.c.Tr.SuggestionsSubtitle,
+							self.c.UserConfig.Keybinding.Universal.Remove, self.c.UserConfig.Keybinding.Universal.Edit)
+					}
```
@jesseduffield I'm showing the keybindings when the suggestions panel gets the focus (and remove it again in the opposite handler in SuggestionsController). I would have preferred not to do this here, but in GetOnFocus/GetOnFocusLost, so that it is reliably called no matter how we switch focus (maybe we want to make it possible some day to click on a suggestion to select it).

But that's not possible right now because `ContextMgr.Replace` only calls Activate on the new context, but not deactivate on the old one. I tried to "fix" that, but that's not possible because the GetOnFocusLost of both ConfirmationController and SuggestionsController destroys the panel.

I'm not sure how to solve this cleanly; let me know if you have any suggestions. For now we just put it here in the tab handler, which works fine.

jesseduffield replied:
Yeah I am not sure how to solve that cleanly. Happy to use your stopgap solution

jesseduffield reviewed (approved):
LGTM
